pkg/controller/istiocsr/certificates_test.go

```diff
@@ -0,0 +1,247 @@
+						istiocsr.DeepCopyInto(o)
+					}
+					return nil
+				})
+			},
+		},
+		{
+			name: "reconciliation of certificate when certificate duration not configured",
+			preReq: func(r *Reconciler, m *fakes.FakeCtrlClient) {
+				m.GetCalls(func(ctx context.Context, ns types.NamespacedName, obj client.Object) error {
+					switch o := obj.(type) {
+					case *v1alpha1.IstioCSR:
+						istiocsr := testIstioCSR(t)
+						istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.CertificateDuration = nil
+						istiocsr.DeepCopyInto(o)
+					}
+					return nil
+				})
+			},
+		},
+		{
+			name: "reconciliation of certificate when certificate RenewBefore not configured",
+			preReq: func(r *Reconciler, m *fakes.FakeCtrlClient) {
+				m.GetCalls(func(ctx context.Context, ns types.NamespacedName, obj client.Object) error {
+					switch o := obj.(type) {
+					case *v1alpha1.IstioCSR:
+						istiocsr := testIstioCSR(t)
+						istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.CertificateRenewBefore = nil
+						istiocsr.DeepCopyInto(o)
+					}
+					return nil
+				})
+			},
+		},
+		{
+			name: "reconciliation of certificate when certificate SignatureAlgorithm not configured",
+			preReq: func(r *Reconciler, m *fakes.FakeCtrlClient) {
+				m.GetCalls(func(ctx context.Context, ns types.NamespacedName, obj client.Object) error {
+					switch o := obj.(type) {
+					case *v1alpha1.IstioCSR:
+						istiocsr := testIstioCSR(t)
+						istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.SignatureAlgorithm = ""
+						istiocsr.DeepCopyInto(o)
+					}
+					return nil
+				})
+			},
+		},
+		{
+			name: "reconciliation of certificate when certificate PrivateKeySize not configured",
+			preReq: func(r *Reconciler, m *fakes.FakeCtrlClient) {
+				m.GetCalls(func(ctx context.Context, ns types.NamespacedName, obj client.Object) error {
+					switch o := obj.(type) {
+					case *v1alpha1.IstioCSR:
+						istiocsr := testIstioCSR(t)
+						istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.PrivateKeySize = 0
+						istiocsr.DeepCopyInto(o)
+					}
+					return nil
+				})
+			},
+		},
+		{
+			name: "reconciliation of certificate when certificate PrivateKeySize and SignatureAlgorithm is misconfigured",
+			preReq: func(r *Reconciler, m *fakes.FakeCtrlClient) {
+				m.GetCalls(func(ctx context.Context, ns types.NamespacedName, obj client.Object) error {
+					switch o := obj.(type) {
+					case *v1alpha1.IstioCSR:
+						istiocsr := testIstioCSR(t)
+						istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.PrivateKeySize = 2048
+						istiocsr.Spec.IstioCSRConfig.IstiodTLSConfig.SignatureAlgorithm = "ECDSA"
+						istiocsr.DeepCopyInto(o)
+					}
+					return nil
+				})
+			},
+			wantErr: `failed to generate certificate resource for creation in istiocsr-test-ns: failed to update certificate resource for istiocsr-test-ns/istiocsr-test-resource istiocsr deployment: certificate parameters PrivateKeySize(2048) and SignatureAlgorithm(ECDSA) do not comply`,
+		},
```

Comment:
@bharath-b-rh Can you please add a testcase where all fields of IstiodTLSConfig are assigned a valid value?

Reply:
Yeah, let me add a scenario.

Reply:
Actually, the `IstiodTLSConfig` is configured with valid values in default `istiocsr` test object initialized [here](https://github.com/openshift/cert-manager-operator/pull/220/files#diff-e2d33866ed923b6322e99d9fd79d0544c7d1834499b7095fee394206b30997d8R61-R68), so in every scenario its being used, except in negative test scenarios related this config. Please let me know if any specific test needs to be added.

Reply:
Ok. This seems good enough.